### ISSUE TITLE
OCP on CNV on OCP (nested clusters)

### ISSUE THREAD

**Issue by durandom** [operate-first/support#277]
**Is your feature request related to a problem? Please describe.**
To support ephemeral clusters, we'd like to explore deploying OCP clusters on CNV
Ultimately having a cluster bot

**Describe the solution you'd like**
* pin down the minimal requirements 
* is either Hetzner or the zero cluster capable of running this
* how can we declaratively describe and install a cluster
* would ACM be able to deploy a cluster

**Additional context**

* https://issues.redhat.com/browse/CNV-12535
* https://events.summit.redhat.com/widget/redhat/sum21/sessioncatalog/session/1617741619673001M92J

cc @rbo @goern 
@fabiand can be asked for more pointers

**Comment by rbo** [comment]
With OpenShift 4.7 kubevirt/cnv is available as Dev Preview
```
$ openshift-install explain installconfig.platform.kubevirt
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <object>
  Kubevirt is the configuration used when installing on kubevirt.

FIELDS:
.....
```

> * pin down the minimal requirements

I assume three node compact cluster is the smallest cluster we can deploy. 3x 16GB RAM + 4 vCPU's

> * is either Hetzner or the zero cluster capable of running this

I will check Hetzner cluster after final installation/sizing.

> * how can we declaratively describe and install a cluster
> * would ACM be able to deploy a cluster

@fabiand do you know if Hive supports kubevirt/cnv?

**Comment by goern** [comment]
@arminmw FYI

**Comment by fabiand** [comment]
@rbo there is partial support in OCP - it will likely never be complete.
@ravidbro can possibly help to leverage the partial support.

Otherwise the cluster can be rought up with AI or UPI.

**Comment by ArminMW** [comment]
How is this different from Scott White's & Arvin Amirian's [Morgan Stanley's presentation on Red Hat Summit 2021](https://events.summit.redhat.com/widget/redhat/sum21/sessioncatalog/session/1617741619673001M92J)
[OpenShift On OpenShift - nested OpenShift](https://static.rainfocus.com/redhat/sum21/sess/1617741619673001M92J/PresentationDeck/Summit%202021%20-%20Morgan%20Stanley_1619807081441001d9Td.pdf) ?

**Comment by ravidbro** [comment]
What they showed in the summit wasn't using IPI with KubeVirt provider, that effort wasn't released eventually although it mentioned in the installer 'explain'.

**Comment by durandom** [comment]
@ravidbro do you have any guides/docs on how to get a POC installed?
My naive approach would be 

1. create appropriate VMs via CNV
2. use UPI to deploy into the VMs (not sure if that would work with ACM)

But I'm sure the devil's in the detail

**Comment by durandom** [comment]
https://github.com/openshift/hypershift-toolkit

Currently examining the path of using Hypershift with kubevirt provisioner

**Comment by fabiand** [comment]
@durandom @rmohr and others are also looking into this.

There is now also https://github.com/okd-virtualization/release - where we prep a consumable version of kubevirt for OKD. This is something we will be able to leverage in this context.

/remove-lifecycle stale

**Comment by fabiand** [comment]
Can we also rename this issue to "KubeVirt/HCO on OKD"

**Comment by rbo** [comment]
> Can we also rename this issue to "KubeVirt/HCO on OKD"

But we running OpenShift Virtualization on OpenShift, so really the product not the upstream projects.

**Comment by fabiand** [comment]
Oh! I was not aware of this. I thought that it was based on OKD

**Comment by durandom** [comment]
Right now we exclusively use OpenShift - but we're open and welcome OKD as well :)
@rmohr @fabiand if any of you are interested in bringing this to the community we can talk about resources.

**Comment by fabiand** [comment]
If it's OCP then it should be pretty straight forward to simply install CNV, as long as we have some BM workers.

Is there any process to enable an operator?

**Comment by durandom** [comment]
We have CNV installed on the rick cluster on BM workers. As documented on https://github.com/operate-first/hetzner-baremetal-openshift

Do you have links how to install OCP on those CNV vms?

Operators are installed via gitops

**Comment by fabiand** [comment]
You say you have CNV installed?
But then also how to install it? :) I am confused.

If you have OCP with BM workers today, then simply install CNV (it's an operator), we could help with a PR.
If you have OCP in VMs on a BM cluster, then we need to enable nesting on the BM nodes, and simply install CNV on OCP inside the VMs.

WHat siutation do we have today?

**Comment by durandom** [comment]
LOL - maybe things progressed since we started....

> If you have OCP in VMs on a BM cluster, then we need to enable nesting on the BM nodes, and simply install CNV on OCP inside the VMs.

We have CNV installed on Rick https://github.com/operate-first/apps/pull/1101 
I don't think we explicitly enabled 'nesting on BM nodes'

How would you recommend deploying an OCP instance on VMs running on the Rick cluster?
We have RHACM available and ArgoCD.
The goal should be to deploy small clusters via gitops or an API

**Comment by fabiand** [comment]
Ah, now I get you:

1. You have CNV installed on OCP BM.
2. You would like to use CNV to deploy tenant clusters in VMs

For 2 we should indeed wait for the hypershift work to converge

**Comment by durandom** [comment]
@fabiand got it. Do you think "KubeVirt/HCO on OKD" will be available sooner?

**Comment by fabiand** [comment]
Yes and no.
 "KubeVirt/HCO on OKD" is simply the upstream pendant of CNV on OCP - thus not a nested cluster.

**Comment by rbo** [comment]
Meanwhile, Hypershift is available as TechPreview with OpenShift 4.10 / ACM 2.5

![image](https://user-images.githubusercontent.com/36604/155685145-2ba5aac6-7c24-4cbf-bd12-4fe3ccfdf42f.png)
